Update Query documentation (#1799)
aspnet/EntityFramework.Docs: entity-framework/core/querying/complex-query-operators.md

---
title: Complex Query Operators - EF Core
author: smitpatel
ms.date: 09/30/2019
ms.assetid: 2e187a2a-4072-4198-9040-aaad68e424fd
uid: core/querying/complex-query-operators
---
# Complex Query Operators

> Language Integrated Query (LINQ) contains many complex operators which combines multiple data sources or does complex processing. Not all LINQ operators have suitable translation in server side. Sometimes, query written in one form can be translated but in different form doing different computation cannot be translated. This page describes some of the complex operators and their supported variations. In future we may recognized more pattern and add their translation. Important point to remember is database also varies in what they support so a particular query which is translated in SqlServer may not work for Sqlite database.

Reviewer: ```suggestion
Language Integrated Query (LINQ) contains many complex operators which combines multiple data sources or does complex processing. Not all LINQ operators have suitable translations on the server side. Sometimes, query written in one form can be translated but in different form it can't be even if the result is same. This page describes some of the complex operators and their supported variations. In future we may recognize more patterns and add their translations. Important point to remember is that database capabilities also vary, so a particular query which is translated on SqlServer may not work for a Sqlite database.
```

> Language Integrated Query (LINQ) contains many complex operators which combines multiple data sources or does complex processing. Not all LINQ operators have suitable translation in server side. Sometimes, query written in one form can be translated but in different form doing different computation cannot be translated. This page describes some of the complex operators and their supported variations. In future we may recognized more pattern and add their translation. Important point to remember is database also varies in what they support so a particular query which is translated in SqlServer may not work for Sqlite database.

Reviewer 2: Casing should be *SQLite* unless you're referring to our packages or APIs.

> [!TIP]
> You can view this article's [sample](https://github.com/aspnet/EntityFramework.Docs/tree/master/samples/core/Querying) on GitHub.

## Join

> LINQ Join operator allows you to connect two data sources based on the key selector on each source and generates a tuple of values when key matches. This naturally translates to `INNER JOIN` on relational databases. While LINQ Join has outer & inner key selectors, databases take a condition as join condition and EF Core will generate join condition by comparing outer key selector to inner key selector through equality operation. Further, if key selectors are anonymous types, EF Core will generate join condition to perform equality component wise.

Reviewer 2: Would *left* and *right* be clearer than *outer* and *inner*?

Reviewer 2: *source* might also be confusing. Maybe *set* is more idiomatic for EF users

Author: Open to ideas. APIs call them outer & inner source.

[!code-csharp[Main](../../../samples/core/Querying/ComplexQuery/Sample.cs#Join)]

```SQL
SELECT [p].[PersonId], [p].[Name], [p].[PhotoId], [p0].[PersonPhotoId], [p0].[Caption], [p0].[Photo]
FROM [PersonPhoto] AS [p0]
INNER JOIN [Person] AS [p] ON [p0].[PersonPhotoId] = [p].[PhotoId]
```

## GroupJoin

> LINQ GroupJoin operator allows you to connect two data sources in same way as Join but the different here is, it creates a group of values from inner source which matches key with element from outer source. Performing query like following generates result of `Blog` & `IEnumerable<Post>`. Since databases (especially relational databases) do not support representing a collection of client side objects, GroupJoin cannot be translated to server in many cases. It requires to get all the data from server to perform GroupJoin without special selector. But if the selector is limiting data being selected then fetching all data from server may cause performance issues. Hence EF Core does not translate GroupJoin.

Reviewer: ```suggestion
LINQ GroupJoin operator allows you to connect two data sources in same way as Join but the difference here is that it creates a group of values from inner source which matches the key with an element from the outer source. Performing a query like the following generates result of `Blog` & `IEnumerable<Post>`. Since databases (especially relational databases) do not support representing a collection of client side objects, GroupJoin cannot be translated to server in many cases. It requires to get all the data from server to perform GroupJoin without a special selector. But if the selector is limiting data being selected then fetching all data from server may cause performance issues. Hence EF Core does not translate GroupJoin.
```

[!code-csharp[Main](../../../samples/core/Querying/ComplexQuery/Sample.cs#GroupJoin)]

[!code-csharp[Main](../../../samples/core/Querying/ComplexQuery/Sample.cs#GroupJoinComposed)]

## SelectMany

> LINQ SelectMany operator allows you to enumerate over collection selector for each element in outer source and generate tuples of value from each data source. In a way it is a join but without any condition so every record is connected with other. Depending on how collection selector is related to outer data source, SelectMany can translate to various different queries on server side.

Reviewer: ```suggestion
LINQ SelectMany operator allows you to enumerate over collection selector for each element in outer source and generate tuples of value from each data source. In a way it is a join but without any condition so every record is connected with another. Depending on how collection selector is related to outer data source, SelectMany can translate to various different queries on server side.
```

### Collection selector does not reference outer

> When collection selector is not referencing anything from outer source, it is cartesian product of both data sources. In translates to `CROSS JOIN` in relational databases.

Reviewer: ```suggestion
When collection selector is not referencing anything from the outer source, it is a cartesian product of both data sources. In translates to `CROSS JOIN` in relational databases.
```

[!code-csharp[Main](../../../samples/core/Querying/ComplexQuery/Sample.cs#SelectManyConvertedToCrossJoin)]

```SQL
SELECT [b].[BlogId], [b].[OwnerId], [b].[Rating], [b].[Url], [p].[PostId], [p].[AuthorId], [p].[BlogId], [p].[Content], [p].[Rating], [p].[Title]
FROM [Blogs] AS [b]
CROSS JOIN [Posts] AS [p]
```

> ### Collection selector references outer in predicate

Reviewer 2: Another word that might be too technical: *predicate*

Author: where? filter?

Reviewer 2: where clause?

> When collection selector has a where predicate which is using element from outer source, normally this arises when using collection navigation on outer source as collection selector, then such where predicate can be used as join condition and it can translate to join on databases. If the collection is empty for outer element, then it would not generate any result for that outer element. But if `DefaultIfEmpty` is applied on collection selector then outer element will be connected with default value of inner element. Due to this distinction, this kind of queries translate to `INNER JOIN` in the absence of `DefaultIfEmpty` and `LEFT JOIN` when `DefaultIfEmpty` is applied.

Reviewer: ```suggestion
When collection selector has a where predicate which is using an element from the outer source (normally this arises when using collection navigation on outer source as collection selector) then such where predicate can be used as the join condition and it can be translated to a join on the server side. If the collection is empty for the outer element, then it would not generate any result for that outer element. But if `DefaultIfEmpty` is applied on collection selector then outer element will be connected with the default value of inner element. Due to this distinction, this kind of queries translate to `INNER JOIN` in the absence of `DefaultIfEmpty` and `LEFT JOIN` when `DefaultIfEmpty` is applied.
```

---
title: Complex Query Operators - EF Core
author: smitpatel
ms.date: 10/03/2019
ms.assetid: 2e187a2a-4072-4198-9040-aaad68e424fd
uid: core/querying/complex-query-operators
---

> # Complex Query Operators

Reviewer 3: We probably need a note that this page applies only to relational database.

Author: That is incorrect.

Reviewer 3: Care to elaborate?

Author: The page is not necessarily put here for relational. We can add more operators here as we go. Like Cosmos can translate bunch of SelectMany in awesome way. The main purpose in not putting relational here is to reuse this page as needed for other providers in future. Further, all the operators talk about the relational translation so it is evident that it applies to relational only.

Reviewer 3: Re Cosmos translations specifically, we may want to consider talking about that in the Cosmos provider docs - not sure it's a good idea to pack all the provider-specific info in one page. It will not be relevant for most readers of this page. Otherwise the page as it is seems 100% relational to me, but I don't have super strong feelings on this.

Reviewer 3: Yeah, there is one point on this long and elaborate page that is also common to non-relational providers.

As I said if you feel very strongly about not including a note on this, I definitely won't die on that hill.

> Language Integrated Query (LINQ) contains many complex operators, which combine multiple data sources or does complex processing. Not all LINQ operators have suitable translation in the server side. Sometimes, query in one form translates to server but if written in different form doesn't translate even if the result is same. This page describes some of the complex operators and their supported variations. In future, we may recognize more pattern and add their translations. Important point to remember is that database also varies in what they support. A particular query, which is translated in SqlServer, mayn't work for SQLite database.

Author: It changes meaning of what we want to say.

> [!TIP]
> You can view this article's [sample](https://github.com/aspnet/EntityFramework.Docs/tree/master/samples/core/Querying) on GitHub.

## Join

> LINQ Join operator allows you to connect two data sources based on the key selector on each source and generates a tuple of values when key matches. It naturally translates to `INNER JOIN` on relational databases. While LINQ Join has outer & inner key selectors, databases take a condition as join condition. So EF Core will generate join condition by comparing outer key selector to inner key selector through equality operation. Further, if key selectors are anonymous types, EF Core will generate join condition to do equality component wise.

Author: Not `The` before LINQ.

[!code-csharp[Main](../../../samples/core/Querying/ComplexQuery/Sample.cs#Join)]

```SQL
SELECT [p].[PersonId], [p].[Name], [p].[PhotoId], [p0].[PersonPhotoId], [p0].[Caption], [p0].[Photo]
FROM [PersonPhoto] AS [p0]
INNER JOIN [Person] AS [p] ON [p0].[PersonPhotoId] = [p].[PhotoId]
```

> ## GroupJoin

Author: No variant of GroupJoin is translatable. Period.
The 2 queries are to demonstrate different amount of data needed to be fetch from server.
There is very little confusion about distinction between join/groupjoin syntax in customers. Most customers so far have known difference. The other query is not `from.... where...` it is actual group join and then composition afterwards. So I don't see any clarification required.

> LINQ GroupJoin operator allows you to connect two data sources in similar to Join but it creates a group of inner values for matching outer elements. Executing query like following generates result of `Blog` & `IEnumerable<Post>`. Since databases (especially relational databases) don't support representing a collection of client-side objects, GroupJoin can't be translated to server in many cases. It requires to get all the data from server to do GroupJoin without special selector (first query below). But if the selector is limiting data being selected then fetching all data from server may cause performance issues (second query below). That's why EF Core doesn't translate GroupJoin.

Reviewer 2: ```suggestion
LINQ's GroupJoin operator allows you to connect two data sources similar to Join, but it creates a group of inner values for matching outer elements. Executing a query like the following generates a result of `Blog` & `IEnumerable<Post>`. Since databases--especially relational databases--don't have a way to represent collections of client-side objects, GroupJoin can't be translated to the server in many cases. It requires you to get all of the data from the server to do GroupJoin without a special selector (first query below). But if the selector is limiting data being selected then fetching all of the data from the server may cause performance issues (second query below). That's why EF Core doesn't translate GroupJoin.
```

[!code-csharp[Main](../../../samples/core/Querying/ComplexQuery/Sample.cs#GroupJoin)]

[!code-csharp[Main](../../../samples/core/Querying/ComplexQuery/Sample.cs#GroupJoinComposed)]

## SelectMany

> LINQ SelectMany operator allows you to enumerate over a collection selector for each outer element and generate tuples of value from each data source. In a way, it's a join but without any condition so every outer element is connected with element from collection source. Depending on how the collection selector is related to the outer data source, SelectMany can translate to various different queries on the server side.

Reviewer 2: ```suggestion
LINQ's SelectMany operator allows you to enumerate over a collection selector for each outer element and generate tuples of values from each data source. In a way, it's a join but without any condition so every outer element is connected with an element from the collection source. Depending on how the collection selector is related to the outer data source, SelectMany can translate into various different queries on the server side.
```

### Collection selector doesn't reference outer

When collection selector isn't referencing anything from the outer source, the result is a cartesian product of both data sources. It translates to `CROSS JOIN` in relational databases.

> [!code-csharp[Main](../../../samples/core/Querying/ComplexQuery/Sample.cs#SelectManyConvertedToCrossJoin)]

Author: Decided not to do. LINQ has both syntax for various thing and LINQ docs can describe that. Our docs would be really long with very low content relating to us.

```SQL
SELECT [b].[BlogId], [b].[OwnerId], [b].[Rating], [b].[Url], [p].[PostId], [p].[AuthorId], [p].[BlogId], [p].[Content], [p].[Rating], [p].[Title]
FROM [Blogs] AS [b]
CROSS JOIN [Posts] AS [p]
```

> ### Collection selector references outer in where clause

Reviewer 2: ```suggestion
### Collection selector references outer in a where clause
```

> When collection selector has a where predicate, which is referencing the outer element, then EF Core translates it to join on database and uses the predicate as join condition. Normally it arises when using collection navigation on outer source as collection selector. If the collection is empty for outer element, then it wouldn't generate any result for that outer element. But if `DefaultIfEmpty` is applied on collection selector then outer element will be connected with default value of inner element. Because of this distinction, this kind of queries translates to `INNER JOIN` in the absence of `DefaultIfEmpty` and `LEFT JOIN` when `DefaultIfEmpty` is applied.

Reviewer 3: ```suggestion
When the collection selector has a Where predicate which references the outer element, EF Core translates it to a database join and uses the predicate as the join condition. Normally this occurs when using a collection navigation on an outer source as the collection selector. If the collection is empty for the outer element, then no results would be generated for that outer element. But if `DefaultIfEmpty` is applied on the collection selector, the outer element will be connected with the default value of the inner element type. Because of this distinction, this kind of queries translates to `INNER JOIN` in the absence of `DefaultIfEmpty` and `LEFT JOIN` when `DefaultIfEmpty` is applied.
```

Author: Acrolinx disallows uses of this.

Reviewer 3: That is not a valid response to a review comment. This fragment is clearly missing articles.

> When the collection selector has a where clause, which references the outer element, then EF Core translates it to a database join and uses the predicate as the join condition. Normally it arises when using collection navigation on the outer element as the collection selector. If the collection is empty for outer element, then no results would be generated for that outer element. But if `DefaultIfEmpty` is applied on the collection selector then the outer element will be connected with default value of the inner element. Because of this distinction, this kind of queries translates to `INNER JOIN` in the absence of `DefaultIfEmpty` and `LEFT JOIN` when `DefaultIfEmpty` is applied.

Reviewer 2: ```suggestion
When the collection selector has a where clause, which references the outer element, then EF Core translates it to a database join and uses the predicate as the join condition. Normally this arises when using collection navigation on the outer element as the collection selector. If the collection is empty for an outer element, then no results would be generated for that outer element. But if `DefaultIfEmpty` is applied on the collection selector then the outer element will be connected with a default value of the inner element. Because of this distinction, this kind of queries translates to `INNER JOIN` in the absence of `DefaultIfEmpty` and `LEFT JOIN` when `DefaultIfEmpty` is applied.
```

[!code-csharp[Main](../../../samples/core/Querying/ComplexQuery/Sample.cs#SelectManyConvertedToJoin)]

```SQL
SELECT [b].[BlogId], [b].[OwnerId], [b].[Rating], [b].[Url], [p].[PostId], [p].[AuthorId], [p].[BlogId], [p].[Content], [p].[Rating], [p].[Title]
FROM [Blogs] AS [b]
INNER JOIN [Posts] AS [p] ON [b].[BlogId] = [p].[BlogId]

SELECT [b].[BlogId], [b].[OwnerId], [b].[Rating], [b].[Url], [p].[PostId], [p].[AuthorId], [p].[BlogId], [p].[Content], [p].[Rating], [p].[Title]
FROM [Blogs] AS [b]
LEFT JOIN [Posts] AS [p] ON [b].[BlogId] = [p].[BlogId]
```

> ### Collection selector references outer in non-where case

Reviewer 2: ```suggestion
### Collection selector references outer in a non-where case
```

> When collection selector has reference to outer, which isn't in predicate (as the above case), it can't be converted to join. That's why we need to evaluate collection selector for each outer element. It translates to `APPLY` operation in relational databases. If the collection is empty for outer element, then it wouldn't generate any result for that outer element. But if `DefaultIfEmpty` is applied on collection selector then outer element will be connected with default value of inner element. Because of this distinction, this kind of queries translates to `CROSS APPLY` in the absence of `DefaultIfEmpty` and `OUTER APPLY` when `DefaultIfEmpty` is applied. Sqlite doesn't support `APPLY` operator.

Reviewer 3: This is quite SQL Server-specific.

```suggestion
When the collection selector references the outer element, but from a non-predicate operator (as was the  case above), the query can't be translated to a regular database join, since we need to evaluate collection selector for each outer element. However, the query does translate to an `APPLY` or a lateral join operation in relational databases.

As before, if the collection is empty for the outer element, then it wouldn't generate any result for that outer element. But if `DefaultIfEmpty` is applied on collection selector then outer element will be connected with the default value of inner element type. Because of this distinction, on SQL Server this kind of query translates to `CROSS APPLY` in the absence of `DefaultIfEmpty` and `OUTER APPLY` when `DefaultIfEmpty` is applied. 

Sqlite doesn't support `APPLY` operator, and therefore this kind of query cannot be translated when targeting it.
```

Author: Lateral join and apply operations are different. It is fine the way it is. It is Oracle-specific.

Reviewer 3: They're not really that different. PostgreSQL implements the APPLY behavior with a LATERAL JOIN, there's just an additional ON TRUE there.

Author: Again Lateral joins are apply operations are different. Postgre fits apply operation on lateral join with additional condition. I have added statement that query may not translate always from EF docs perspective, we generate Apply operation and that is accurate doc content.

Reviewer 3: The fact is that what this section describes (Collection selector references outer in non-predicate) is translated in PostgreSQL via lateral join, which is part of the SQL standard; PostgreSQL users reading this will have no idea about APPLY. All I'm asking is to mention lateral join alongside APPLY so that they understand, because that's what is used for translation.

Author: That is misleading. Relational generates Apply operation. EF Core relational layer does not generate lateral joins. Writing lateral join easily confuses user asking, oh why did not translate this query via lateral join. Sorry but we don't support lateral join.

Reviewer 3: I think I'll let this PR go now.

> When the collection selector has reference to the outer element, which isn't in where clause as the case above, it doesn't translate to database join. That's why we need to evaluate the collection selector for each outer element. It translates to `APPLY` operation in many relational databases. If the collection is empty for outer element, then no results would be generated for that outer element. But if `DefaultIfEmpty` is applied on the collection selector then the outer element will be connected with default value of the inner element. Because of this distinction, this kind of queries translates to `CROSS APPLY` in the absence of `DefaultIfEmpty` and `OUTER APPLY` when `DefaultIfEmpty` is applied. Certain databases like SQLite don't support `APPLY` operator so this kind of query may not be translated.

Reviewer 2: ```suggestion
When the collection selector references the outer element, which isn't in a where clause (as the case above), it doesn't translate to a database join. That's why we need to evaluate the collection selector for each outer element. It translates to `APPLY` operations in many relational databases. If the collection is empty for an outer element, then no results would be generated for that outer element. But if `DefaultIfEmpty` is applied on the collection selector then the outer element will be connected with a default value of the inner element. Because of this distinction, these kinds of queries translate to `CROSS APPLY` in the absence of `DefaultIfEmpty` and `OUTER APPLY` when `DefaultIfEmpty` is applied. Certain databases like SQLite don't support `APPLY` operators so this kind of query may not be translated.
```

[!code-csharp[Main](../../../samples/core/Querying/ComplexQuery/Sample.cs#SelectManyConvertedToApply)]

```SQL
SELECT [b].[BlogId], [b].[OwnerId], [b].[Rating], [b].[Url], ([b].[Url] + N'=>') + [p].[Title] AS [p]
FROM [Blogs] AS [b]
CROSS APPLY [Posts] AS [p]

SELECT [b].[BlogId], [b].[OwnerId], [b].[Rating], [b].[Url], ([b].[Url] + N'=>') + [p].[Title] AS [p]
FROM [Blogs] AS [b]
OUTER APPLY [Posts] AS [p]
```

## GroupBy

> LINQ GroupBy operator creates result of type `IGrouping<TKey, TElement>` where `TKey` & `TElement` could be arbitrary types. Further, `IGrouping` implements `IEnumerable<TElement>`, which means you can compose over using any LINQ operator after grouping. Since no database structure can represent IGrouping, Group By operator has no translation in most cases. When an aggregate operator is applied to each group to which gives back a scalar, it can be translated to SQL `GROUP BY` in relational databases. SQL `GROUP BY` is restrictive too. It requires you to group by only scalar values  projection can contain only grouping key columns or aggregate applied over any column. EF Core identifies this subset and translates it to server. Following example shows a query, which has aggregate operator applied to it and generated SQL.

Reviewer 3: ```suggestion
The LINQ GroupBy operator returns results of type `IGrouping<TKey, TElement>`, where `TKey` and `TElement` could be arbitrary types. Further, `IGrouping` implements `IEnumerable<TElement>`, which means it can be composed over using any LINQ operator after grouping. Since no database structure can represent `IGrouping`, the GroupBy operator cannot be translated in most cases. As an exception, when an aggregate operator is applied to each group (e.g. MAX, COUNT), the query can be translated to  SQL `GROUP BY`. 

Note that SQL `GROUP BY` is restrictive too, requiring you to group only by scalar values; it also requires the projection to only contain grouping key columns or applied aggregate operators. EF Core identifies this subset and translates it to server, as in the following example:
```

Author: Acrolinx doesn't like long statements.

Author: Acrolinx believe that Note that does not add any value and unncessary.

> LINQ GroupBy operator creates result of type `IGrouping<TKey, TElement>` where `TKey` and `TElement` could be arbitrary types. Further, `IGrouping` implements `IEnumerable<TElement>`, which means you can compose over using any LINQ operator after grouping. Since no database structure can represent `IGrouping`, Group By operator has no translation in most cases. When an aggregate operator is applied to each group, which gives back a scalar, it can be translated to SQL `GROUP BY` in relational databases. SQL `GROUP BY` is restrictive too. It requires you to group by only scalar values  projection can contain only grouping key columns or aggregate applied over any column. EF Core identifies this subset and translates it to server, as in the following example:

Reviewer 2: ```suggestion
LINQ GroupBy operators create a result of type `IGrouping<TKey, TElement>` where `TKey` and `TElement` could be any arbitrary type. Furthermore, `IGrouping` implements `IEnumerable<TElement>`, which means you can compose over it using any LINQ operator after the grouping. Since no database structure can represent an `IGrouping`, Group By operator has no translation in most cases. When an aggregate operator is applied to each group, which returns a scalar, it can be translated to SQL `GROUP BY` in relational databases. SQL's `GROUP BY` is restrictive too. It requires you to group only by scalar values. The projection can only contain grouping key columns or any aggregate applied over a column. EF Core identifies this pattern and translates it to the server, as in the following example:
```

[!code-csharp[Main](../../../samples/core/Querying/ComplexQuery/Sample.cs#GroupBy)]

```SQL
SELECT [p].[AuthorId] AS [Key], COUNT(*) AS [Count]
FROM [Posts] AS [p]
GROUP BY [p].[AuthorId]
```

> EF Core also translates when aggregate operator appears in Where filter by using `HAVING` clause in SQL and aggregate operator appears in Order By (or other ordering operators). The query before applying Group By operator can be any complex query using various different operators as long as it can be translated to server. Further, once you apply aggregate operator on grouping query to remove groupings from resulting source, you can compose it like any other query.

Reviewer 3: ```suggestion
EF Core can also translate queries where an aggregate operator appears in a Where filter (by generating an SQL `HAVING` clause), or in an OrderBy clause (or other ordering operators). The query part before the GroupBy operator can be any complex query as long as it can be translated to server. Further, once you apply aggregate operator on a grouping query as shown above, you can compose it like any other query.
```

Author: Avoid usage of can when we do it.

Author: Acrolinx also dislikes parenthesis.

> EF Core also translates queries where an aggregate operator on the grouping appears in Where or OrderBy (or other ordering) LINQ operators. It uses `HAVING` clause in SQL for where clause. The query part before applying GroupBy operator can be any complex query as long as it can be translated to server. Further, once you apply aggregate operators on a grouping query to remove groupings from resulting source, you can compose it like any other query.

Reviewer 2: ```suggestion
EF Core also translates queries where an aggregate operator on the grouping appears in a Where or OrderBy (or other ordering) LINQ operator. It uses `HAVING` clauses in SQL for the where clause. The part of the query before applying the GroupBy operator can be any complex query as long as it can be translated to server. Furthermore, once you apply aggregate operators on a grouping query to remove groupings from the resulting source, you can compose on top of it like any other query.
```

[!code-csharp[Main](../../../samples/core/Querying/ComplexQuery/Sample.cs#GroupByFilter)]

```SQL
SELECT [p].[AuthorId] AS [Key], COUNT(*) AS [Count]
FROM [Posts] AS [p]
GROUP BY [p].[AuthorId]
HAVING COUNT(*) > 0
ORDER BY [p].[AuthorId]
```

The aggregate operators EF Core supports are following

- Average
- Count
- LongCount
- Max
- Min
- Sum

> ## Left Join

Reviewer 3: This repeats the above GroupJoin section no?

Author: Nope.

> The aggregate operators EF Core supports are following

Reviewer 2: ```suggestion
The aggregate operators EF Core supports are as follows
```

> While Left Join isn't a LINQ operator, relational database has concept of Left Join which is frequently used in queries. A particular pattern in LINQ query gives same result as `LEFT JOIN` on server. EF Core identifies such pattern and generate equivalent `LEFT JOIN` on server side. The pattern involves creating GroupJoin between both the data sources and then flattening out the grouping by using SelectMany operator with DefaultIfEmpty on the grouping source to match null when inner doesn't have related element. Following example shows how that pattern looks like what it generates.

Reviewer 2: ```suggestion
While Left Join isn't a LINQ operator, relational databases have the concept of a Left Join which is frequently used in queries. A particular pattern in LINQ queries gives the same result as a `LEFT JOIN` on the server. EF Core identifies such patterns and generates the equivalent `LEFT JOIN` on the server side. The pattern involves creating a GroupJoin between both the data sources and then flattening out the grouping by using the SelectMany operator with DefaultIfEmpty on the grouping source to match null when the inner doesn't have a related element. The following example shows what that pattern looks like and what it generates.
```

[!code-csharp[Main](../../../samples/core/Querying/ComplexQuery/Sample.cs#LeftJoin)]

```SQL
SELECT [b].[BlogId], [b].[OwnerId], [b].[Rating], [b].[Url], [p].[PostId], [p].[AuthorId], [p].[BlogId], [p].[Content], [p].[Rating], [p].[Title]
FROM [Blogs] AS [b]
LEFT JOIN [Posts] AS [p] ON [b].[BlogId] = [p].[BlogId]
```

> Above pattern creates complex structure in expression tree. Because of that EF Core requires you to flatten out grouping results out of GroupJoin operator in immediate step. Even if the GroupJoin-DefaultIfEmpty-SelectMany is used but in different pattern, we may not identify as Left Join.

Reviewer 2: ```suggestion
The above pattern creates a complex structure in the expression tree. Because of that, EF Core requires you to flatten out the grouping results of the GroupJoin operator in a step immediately following the operator. Even if the GroupJoin-DefaultIfEmpty-SelectMany is used but in a different pattern, we may not identify it as a Left Join.
```
